### PR TITLE
[WIP] Flp subtimeframe chain

### PR DESCRIPTION
**Pull request by ironMann**
FLP chain: Building SubTimeFrames consisting of readout data
using FairMQUnmanagedRegion

Components:
  - Readout O2 Device (with a CRU emulator)
  - SubTimeFrameBuilder O2 Device
  - SubTimeFrameTransporter O2 Device

STF Serialization options:
  1) header and data interleaved: each logical/physical data
     block is preceded by a header.
  2) headers first: all headers are sent first, then all
     physical data blocks.

### CODE REVIEW

File: Utilities/DataDistribution/Common/include/Common/ConcurrentQueue.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+namespace impl {
+
+/// Concurrent (thread-safe) container adapter for FIFO/LIFO data structures
+enum QueueType {
+  eFIFO,
+  eLIFO
+};
+
+template <typename T, QueueType type>
+class ConcurrentContainerImpl {
```
Comment (sawenzel) [inline Utilities/DataDistribution/Common/include/Common/ConcurrentQueue.h]: just for my understanding: Is there nothing available in C++11 or in boost?

File: Utilities/DataDistribution/Common/include/Common/O2ExtentVector.h [hunk 2]
```diff
@@ -0,0 +1,1110 @@
...
+
+namespace detail {
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+// ExtentPosition
+//
+////////////////////////////////////////////////////////////////////////////////////////////////////
+template <class Allocator>
+class Extent {
```
Comment (sawenzel) [inline Utilities/DataDistribution/Common/include/Common/O2ExtentVector.h]: could you add reasonable descriptions/explanations to new classes? What are they doing?

**Comment by sawenzel** [comment]
@ironMann : Could you try adding some documentation to this new code? I am personally missing explanations for new classes. Can you for instance try to clarify what is special about the newly introduced containers? Why can't one use std:: containers or solutions from boost? Such explanations would also enhance the understanding of what is being done here. Thanks

**Comment by ironMann** [comment]
@sawenzel Sorry for the delay, I'm on the road until next week. I'll try to address the questions regarding DataDistribution/Common classes. I personally do not know any direct replacements from other libs... any suggestion is appreciated.

I'll drop the `O2ExtentVector` class from this PR, which is not used (for now). Some of the motivation for it is in [this document](https://paper.dropbox.com/doc/O2ExtentVector-tyybR2n2kHRoFUlounfKb?_tk=share_copylink). This needs a lot of work, but I think the need exists.

**Comment by shahor02** [comment]
@ironMann : as far as I understand, the O2ExtentVector is the class which was supposed to substitute my old PR for copyable vector? I confirm it is needed and it will be used once available (even if with limited functionality).

**Comment by sawenzel** [comment]
@ironMann @shahor02 : Thanks. Clearly having documentation on classes helps in understanding their purpose.
As for O2ExtendVector, I am wondering if `std::deque` does not already do what is needed?

**Comment by sawenzel** [comment]
More 5cents on the adopt requirement: We already have the guideline support library which provides `gsl::span` (or array_view). This thing takes an existing memory buffer and provides a vector view on top ... which sounds like the thing that you want to do. In other words, a sender device could send a std::vector (as a buffer) and the receiving device could consume the vector using a view.

File: Utilities/DataDistribution/Common/include/Common/ReadoutDataModel.h [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  // might as well piggyback this info to avoid dedicated message
+  std::uint64_t mStfId;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+/// CRU Link
+////////////////////////////////////////////////////////////////////////////////
+
+// This data struct is transmitted 'atomically' by the Readout
+struct O2CruLinkHeader : public O2DataHeader {
```
Comment (matthiasrichter) [inline Utilities/DataDistribution/Common/include/Common/ReadoutDataModel.h]: `O2CRULinkHeader` must inherit from o2::Header::BaseHeader`; regarding naming, the O2 prefix should be removed as we use namespace o2

**Comment by mkrzewic** [comment]
@sawenzel, indeed the functionality is there with array_view - so we can use it, but it would introduce a bit of an asymmetry to have a different abstraction for writing and reading. Don't have a strong opinion here.

@matthiasrichter regarding the inheritance from BaseHeader: in fact some people pointed out (Gvozden) that there is no layout guarantee in the standard for inheritance... So formally we should switch to composition. Another potential problem is reading back the headers the way we do since there might be unaligned access issues. Question is: do we keep this in the back of our heads and fix it when (and if) this becomes a problem, or do we fix it now? (fix would involve a copy at read, I think in reality we don't have to worry about performance here.)

**Comment by ironMann** [comment]
Hi, sorry about the stall here. The PR in this form should not be merged, but it can serve as a basis for discussion and convergence on standardizing of these components. There are several points that I'll put on the agenda for meeting on Nov 30t: e.g. why the std::vector might be a bad choice in the context of shared memory; how to add some typing information to structures carried in messages; concerns with undefined behavior, etc. Also, I'm available throughout next week for discussions.

**Comment by sawenzel** [comment]
@ironMann : Marked this PR as WIP since it should not be merged.

**Comment by sawenzel** [comment]
What is the status of this PR? Shall it be closed?

**Comment by sawenzel** [comment]
Closing this PR for the moment since no activity. Please reopen as you see fit.

**Pull request closed without merging**